library/std/src/sys/windows/mutex.rs

```diff
@@ -37,106 +30,42 @@ pub type MovableMutex = Mutex;
             // initializing an SRWLOCK here.
-            lock: AtomicUsize::new(0),
+            srwlock: UnsafeCell::new(c::SRWLOCK_INIT),
         }
     }
     #[inline]
     pub unsafe fn init(&mut self) {}
+
+    #[inline(always)]
     pub unsafe fn lock(&self) {
-        match kind() {
-            Kind::SRWLock => c::AcquireSRWLockExclusive(raw(self)),
-            Kind::CriticalSection => {
-                let inner = &*self.inner();
-                inner.remutex.lock();
-                if inner.held.replace(true) {
-                    // It was already locked, so we got a recursive lock which we do not want.
-                    inner.remutex.unlock();
-                    panic!("cannot recursively lock a mutex");
-                }
-            }
-        }
+        c::AcquireSRWLockExclusive(raw(self));
     }
+
+    #[inline(always)]
     pub unsafe fn try_lock(&self) -> bool {
-        match kind() {
-            Kind::SRWLock => c::TryAcquireSRWLockExclusive(raw(self)) != 0,
-            Kind::CriticalSection => {
-                let inner = &*self.inner();
-                if !inner.remutex.try_lock() {
-                    false
-                } else if inner.held.replace(true) {
-                    // It was already locked, so we got a recursive lock which we do not want.
-                    inner.remutex.unlock();
-                    false
-                } else {
-                    true
-                }
-            }
-        }
+        c::TryAcquireSRWLockExclusive(raw(self)) != 0
     }
+
+    #[inline(always)]
     pub unsafe fn unlock(&self) {
-        match kind() {
-            Kind::SRWLock => c::ReleaseSRWLockExclusive(raw(self)),
-            Kind::CriticalSection => {
-                let inner = &*(self.lock.load(Ordering::SeqCst) as *const Inner);
-                inner.held.set(false);
-                inner.remutex.unlock();
-            }
-        }
-    }
-    pub unsafe fn destroy(&self) {
-        match kind() {
-            Kind::SRWLock => {}
-            Kind::CriticalSection => match self.lock.load(Ordering::SeqCst) {
-                0 => {}
-                n => Box::from_raw(n as *mut Inner).remutex.destroy(),
-            },
-        }
+        c::ReleaseSRWLockExclusive(raw(self));
     }
 
-    unsafe fn inner(&self) -> *const Inner {
-        match self.lock.load(Ordering::SeqCst) {
-            0 => {}
-            n => return n as *const _,
-        }
-        let inner = box Inner { remutex: ReentrantMutex::uninitialized(), held: Cell::new(false) };
-        inner.remutex.init();
-        let inner = Box::into_raw(inner);
-        match self.lock.compare_exchange(0, inner as usize, Ordering::SeqCst, Ordering::SeqCst) {
-            Ok(_) => inner,
-            Err(n) => {
-                Box::from_raw(inner).remutex.destroy();
-                n as *const _
-            }
-        }
+    pub unsafe fn destroy(&self) {
```

Comment:
If you're making the rest `#[inline]`, might as well add it here.

```suggestion
    #[inline]
    pub unsafe fn destroy(&self) {
```